experiments/process_sandbox/library_runner.cc

```diff
@@ -239,184 +219,531 @@ namespace sandbox
   void ProxyPageMap::set_large_size(void* p, size_t size)
   {
     size_t size_bits = bits::next_pow2_bits(size);
-    if ((p >= shared_memory_start) && (p < shared_memory_end))
+    SANDBOX_DEBUG_INVARIANT(
+      (p >= shared_memory_start) && (p < shared_memory_end),
+      "Setting large size for pointer {} in pagemap that is outside of the "
+      "sandbox range {}--{}",
+      p,
+      shared_memory_start,
+      shared_memory_end);
+    requestHostService(
+      ChunkMapSetRange,
+      reinterpret_cast<uintptr_t>(p),
+      static_cast<uintptr_t>(size_bits));
+  }
+
+  void ProxyPageMap::clear_large_size(void* p, size_t size)
+  {
+    SANDBOX_DEBUG_INVARIANT(
+      (p >= shared_memory_start) && (p < shared_memory_end),
+      "Clearing large size for pointer {} in pagemap that is outside of the "
+      "sandbox range {}--{}",
+      p,
+      shared_memory_start,
+      shared_memory_end);
+    size_t size_bits = bits::next_pow2_bits(size);
+    requestHostService(
+      ChunkMapClearRange,
+      reinterpret_cast<uintptr_t>(p),
+      static_cast<uintptr_t>(size_bits));
+  }
+
+  void* MemoryProviderProxy::pop_large_stack(size_t large_class)
+  {
+    return reinterpret_cast<void*>(requestHostService(
+      MemoryProviderPopLargeStack, static_cast<uintptr_t>(large_class)));
+  }
+
+  void
+  MemoryProviderProxy::push_large_stack(Largeslab* slab, size_t large_class)
+  {
+    requestHostService(
+      MemoryProviderPushLargeStack,
+      reinterpret_cast<uintptr_t>(slab),
+      static_cast<uintptr_t>(large_class));
+  }
+
+  void* MemoryProviderProxy::reserve_committed_size(size_t size) noexcept
+  {
+    size_t size_bits = snmalloc::bits::next_pow2_bits(size);
+    size_t large_class = std::max(size_bits, SUPERSLAB_BITS) - SUPERSLAB_BITS;
+    return reserve_committed(large_class);
+  }
+  void* MemoryProviderProxy::reserve_committed(size_t large_class) noexcept
+  {
+    return reinterpret_cast<void*>(requestHostService(
+      MemoryProviderReserve, static_cast<uintptr_t>(large_class)));
+  }
+
+}
+
+namespace
+{
+  /**
+   * The function from the loaded library that provides the vtable dispatch
+   * for functions that we invoke.
+   */
+  void (*sandbox_invoke)(int, void*);
+
+  /**
+   * The run loop.  Takes the public interface of this library (effectively,
+   * the library's vtable) as an argument.  Exits when the upcall depth changes
+   * after executing the helper function.  This provides a nested runloop
+   * abstraction similar to OpenStep's modal runloop. Each recursion depth in
+   * an upcall has its own runloop that handles recursive invocations from the
+   * parent in response to the upcall.
```

Comment:
Is this about upcalls that need other upcalls to complete? So, the runloop triggers an upcall, which calls another runloop with `++upcall_depth` as an argument, and when that's done, it comes back to the current runloop, etc.?